src/VecSim/algorithms/brute_force/bf_batch_iterator.cpp

```diff
@@ -0,0 +1,146 @@
+    CandidatesHeap TopCandidates;
+    // map vector's label to its index in the scores vector.
+    unordered_map<size_t, size_t> TopCandidatesIndices(n_res);
+    for (size_t i = this->scores_valid_start_pos; i < this->scores.size(); i++) {
+        if (isnan(this->scores[i].first)) {
+            continue;
+        }
+        if (TopCandidates.size() < n_res) {
+            TopCandidates.emplace(this->scores[i].first, this->scores[i].second);
+            TopCandidatesIndices[this->scores[i].second] = i;
+            upperBound = TopCandidates.top().first;
+        } else {
+            if (this->scores[i].first >= upperBound) {
+                continue;
+            } else {
+                TopCandidates.emplace(this->scores[i].first, this->scores[i].second);
+                TopCandidatesIndices[this->scores[i].second] = i;
+                // remove the furthest vector from the candidates and from the label->index mappings
+                TopCandidatesIndices.erase(TopCandidates.top().second);
+                TopCandidates.pop();
+                upperBound = TopCandidates.top().first;
+            }
+        }
+    }
+
+    auto *results = array_new_len<VecSimQueryResult>(TopCandidates.size(), TopCandidates.size());
+    for (int i = (int)TopCandidates.size() - 1; i >= 0; --i) {
+        VecSimQueryResult_SetId(results[i], TopCandidates.top().second);
+        VecSimQueryResult_SetScore(results[i], TopCandidates.top().first);
+        // Invalidate vector score, so we won't return it again in the next iterations.
+        this->scores[TopCandidatesIndices[TopCandidates.top().second]].first = INVALID_SCORE;
+        TopCandidates.pop();
+    }
+    return results;
+}
+
+VecSimQueryResult *BF_BatchIterator::selectBasedSearch(size_t n_res) {
+    size_t remaining_vectors_count = this->scores.size() - this->scores_valid_start_pos;
+    // Get an iterator to the effective first element in the scores array, which is the first
+    // element that hasn't been returned in previous iterations.
+    auto valid_begin_it = this->scores.begin() + (int)(this->scores_valid_start_pos);
+    // We return up to n_res vectors, the remaining vectors size is an upper bound.
+    if (n_res > remaining_vectors_count) {
+        n_res = remaining_vectors_count;
+    }
+    auto n_th_element_pos = valid_begin_it + (int)n_res;
+    // This will perform an in-place partition of the elements in the slice of the array that
+    // contains valid results, based on the n-th element as the pivot - every element with a lower
+    // will be placed before it, and all the rest will be placed after.
+    std::nth_element(valid_begin_it, n_th_element_pos, this->scores.end(), cmp);
+
+    auto *results = array_new<VecSimQueryResult>(n_res);
+    for (size_t i = this->scores_valid_start_pos; i < this->scores_valid_start_pos + n_res; i++) {
+        if (isnan(this->scores[i].first)) {
+            continue;
+        }
```

Comment:
is this a valid scenario?

Reply:
Made an update on Thursday night that makes this check no longer required (+ improving performance for heap based search)